renderdoc/driver/d3d8/d3d8_resources.h

```diff
@@ -0,0 +1,250 @@
+    m_pDevice->GetResourceManager()->ReleaseCurrentResource(GetResourceID());
+    m_pDevice->ReleaseResource((NestedType *)this);
+    SAFE_RELEASE(m_pReal);
+    m_pDevice = NULL;
+  }
+
+  virtual ~WrappedResource8()
+  {
+    // should have already called shutdown (needs to be called from child class to ensure
+    // vtables are still in place when we call ReleaseResource)
+    RDCASSERT(m_pDevice == NULL && m_pReal == NULL);
+  }
+
+public:
+  typedef NestedType InnerType;
+
+  NestedType *GetReal() { return m_pReal; }
+  ULONG STDMETHODCALLTYPE AddRef() { return RefCounter8::SoftRef(m_pDevice) - m_PipelineRefs; }
+  ULONG STDMETHODCALLTYPE Release()
+  {
+    unsigned int piperefs = m_PipelineRefs;
+    return RefCounter8::SoftRelease(m_pDevice) - piperefs;
+  }
+
+  void PipelineAddRef() { InterlockedIncrement(&m_PipelineRefs); }
+  void PipelineRelease() { InterlockedDecrement(&m_PipelineRefs); }
+  HRESULT STDMETHODCALLTYPE QueryInterface(REFIID riid, void **ppvObject)
+  {
+    if(riid == __uuidof(IUnknown))
+    {
+      *ppvObject = (IUnknown *)(NestedType *)this;
+      AddRef();
+      return S_OK;
+    }
+    if(riid == m_NestedTypeIID)
+    {
+      *ppvObject = (NestedType *)this;
+      AddRef();
+      return S_OK;
+    }
+
+    // DEFINE_GUID(IID_IDirect3DResource8, 0x1b36bb7b, 0x9b7, 0x410a, 0xb4, 0x45, 0x7d, 0x14, 0x30,
+    // 0xd7, 0xb3, 0x3f);
+    static const GUID IDirect3DResource8_uuid = {0x1b36bb7b, 0x9b7, 0x410a, 0xb4, 0x45, 0x7d,
+                                                 0x14,       0x30,  0xd7,   0xb3, 0x3f};
+
+    if(riid == IDirect3DResource8_uuid)
+    {
+      *ppvObject = (IDirect3DResource8 *)this;
+      AddRef();
+      return S_OK;
+    }
+
+    return RefCounter8::QueryInterface(riid, ppvObject);
+  }
+
+  //////////////////////////////
+  // implement IDirect3DResource8
+
+  HRESULT STDMETHODCALLTYPE GetDevice(__out IDirect3DDevice8 **ppDevice)
+  {
+    if(ppDevice)
+    {
+      *ppDevice = m_pDevice;
+      m_pDevice->AddRef();
+      return S_OK;
+    }
+    return E_INVALIDARG;
+  }
+
+  HRESULT STDMETHODCALLTYPE GetPrivateData(REFGUID guid, void *pData, DWORD *pSizeOfData)
+  {
+    return m_pReal->GetPrivateData(guid, pData, pSizeOfData);
+  }
+
+  HRESULT STDMETHODCALLTYPE SetPrivateData(REFGUID guid, const void *pData, DWORD SizeOfData,
+                                           DWORD Flags)
+  {
+    return m_pReal->SetPrivateData(guid, pData, SizeOfData, Flags);
+  }
+
+  HRESULT STDMETHODCALLTYPE FreePrivateData(REFGUID guid) { return m_pReal->FreePrivateData(guid); }
+  DWORD STDMETHODCALLTYPE SetPriority(DWORD PriorityNew)
+  {
+    return m_pReal->SetPriority(PriorityNew);
+  }
+
+  DWORD STDMETHODCALLTYPE GetPriority() { return m_pReal->GetPriority(); }
+  void STDMETHODCALLTYPE PreLoad() { return m_pReal->PreLoad(); }
+  D3DRESOURCETYPE STDMETHODCALLTYPE GetType() { return m_pReal->GetType(); }
+};
+
+template <typename NestedType, typename DescType>
```

Comment:
For what it's worth, on D3D11 I used the `DescType` template parameter to also store a copy of the Desc inside the object in the constructor - in non-release builds only. I've found it very useful for debugging, but it's up to you if you want to do that or not.